### PR TITLE
Reworked caption body checks on components 

### PR DESCRIPTION
**Pull request by Heulitig**
[Old PR](https://github.com/FifthTry/ftd/pull/243)
- Revamped the code on caption body checks (**wip**) 
- This is the continuation of an old PR which is closed now
- Refer issue - https://github.com/FifthTry/ftd/issues/217

# Things added: 
```markdown
# Caption/Body/Header_value conflicts
- This happens if any argument accepts data from more than one way
 and the user doesn't pass this data in exactly one way

# Missing data checks
- This happens if any (required) argument doesn't get the data from any way it takes

# Unknown data checks
- This happens when there is no argument to accept the data passed from caption/body
```
- Tests added under `component.rs`

### CODE REVIEW

File: src/component.rs [hunk 1]
```diff
@@ -600,6 +600,12 @@ impl ChildComponent {
         let root_property =
             get_root_property(line_number, name, caption, doc, &all_arguments, inherits)?;
 
+        // dbg!(root.full_name.as_str());
+        // dbg!(name);
+        // println!("Here 1");
+
+        assert_caption_body_checks(&root.full_name, p1, doc, caption, body, line_number)?;
+
```
Comment (amitu) [inline src/component.rs:608]: Delete these comments please.

**Comment by AbrarNitk** [comment]
@Heulitig did you test it with `fpm`?

**Comment by Heulitig** [comment]
> @Heulitig did you test it with `fpm`?

I have used the local ftd in fpm in testing the other repos like ftd.dev, fpm.dev, doc-site, dotcom, df. And also did check against ftd tests and fpm tests.

**Comment by amitu** [comment]
@Heulitig should I merge this?

**Comment by amitu** [comment]
@Arpita-Jaiswal, @Heulitig, this PR is open since long. Can we merge it?

**Review by Arpita-Jaiswal — CHANGES_REQUESTED** [review]

File: src/component.rs [hunk 2]
```diff
@@ -2617,6 +2637,376 @@ pub fn read_properties(
...
+        root.eq("ftd#ui")
+    }
+
+    /// fetches the thing from the bag
+    fn get_thing<'t>(
+        bag_entry: &str,
+        bag: &'t std::collections::BTreeMap<String, ftd::p2::Thing>,
+        doc_id: &str,
+        line_number: usize,
+    ) -> ftd::p1::Result<&'t ftd::p2::Thing> {
```
[the remark this answers is not in the document]
Comment (Heulitig) [reply src/component.rs:2719]: I didn't knew a similar function existed for this purpose. Using the existing one then and removing this function.

**Comment by amitu** [comment]
@Arpita-Jaiswal should we merge this now?

**Comment by Arpita-Jaiswal** [comment]
> @Arpita-Jaiswal should we merge this now?

Yep

**Pull request merged**